src/git.rs

```diff
@@ -357,3 +376,69 @@ pub fn yank(
         Ok(())
     })
 }
+
+/// Collapse the index into a single commit, archiving the current history in a snapshot branch.
+#[swirl::background_job]
+pub fn squash_index(env: &Environment) -> Result<(), PerformError> {
+    let repo = env.lock_index()?;
+    println!("Squashing the index into a single commit.");
+
+    let now = Utc::now().format("%Y-%m-%d");
+    let original_head = repo.head_oid()?;
+    let msg = format!("Collapse index into one commit\n\n\
+
+        Previous HEAD was {}, now on the `snapshot-{}` branch\n\n\
+
+        More information about this change can be found [online] and on [this issue].\n\n\
+
+        [online]: https://internals.rust-lang.org/t/cargos-crate-index-upcoming-squash-into-one-commit/8440\n\
+        [this issue]: https://github.com/rust-lang/crates-io-cargo-teams/issues/47", original_head, now);
+
+    // Create a snapshot branch of current `HEAD`.
+    repo.push(&format!("HEAD:refs/heads/snapshot-{}", now))?;
+
+    repo.squash_to_single_commit(&msg)?;
+
+    // Shell out to git because libgit2 does not currently support push leases
+
+    let key = match &repo.credentials {
+        Credentials::Ssh { key } => key,
+        Credentials::Http { .. } => {
+            return Err(String::from("squash_index: Password auth not supported").into())
+        }
+        _ => return Err(String::from("squash_index: Could not determine credentials").into()),
+    };
+
+    let temp_key_file = tempfile::Builder::new().tempfile()?;
```

Comment:
On Linux (and the production environment) this will create the file in `/tmp`, but it'd be better to create it in `/dev/shm`: files stored in `/dev/shm` are kept in memory and never persisted on disk. On other OSes we can use the default.

Reply:
Agree, done in first new commit.